[Backtracing] Update module maps.

al45tair opened the pull request:
Now that Darwin is provided by the system, we no longer need to disable the system module maps in favour of our own, which also resolves a few other problems that doing that was creating.

rdar://137201928

al45tair committed:
[Backtracing] Update module maps.

Now that Darwin is provided by the system, we no longer need to disable the
system module maps in favour of our own, which also resolves a few other
problems that doing that was creating.

rdar://137201928

al45tair committed:
[Backtracing] Use C++ interop to fix elf.h issue on Linux.

We have some problems on Linux where Glibc pulls in `<elf.h>` and then
we end up with conflicting definitions.  Fix by using C++ interop and
putting our definitions into a namespace.

rdar://137201928

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test

ian-twilightcoder commented on stdlib/public/Backtracing/Backtrace.swift:
```diff
@@ -16,14 +16,22 @@
 
-@_implementationOnly import OS.Libc
+#if os(macOS) || os(iOS) || os(tvOS) || os(watchOS)
+internal import Darwin
```
Darwin is pretty heavy, is it possible to import one of the modules under it like `errno_h` or `stdio_h` if they're available?

al45tair replied:
Maybe. That might involve importing a number of modules, mind, and it'll only really affect the compile time of the backtracing library (since we aren't exporting anything from them). So it's worth experimenting with, for sure, but we can probably defer that for now.

ian-twilightcoder replied:
I was thinking more like pulling in dependencies you don't want rather than compile time. i.e. do you really want bsm and nurses and all the other bs that Darwin gives you? Or do you just need a few cstdlib things?

ian-twilightcoder reviewed (approved):
🥳

al45tair commented:
Ah. The Linux problem is one I already have fixed elsewhere (by using C++ interop). I might need to pull in some changes from a different PR to fix that.

al45tair commented:
@swift-ci Please test

al45tair commented:
Joy. That's not going to work either. Still more changes required.

al45tair committed:
[Backtracing] Turn off MemberImportVisibility.

Also fixed a couple of other minor issues preventing this PR from
building on its own.

rdar://137201928

al45tair commented:
@swift-ci Please test

al45tair committed:
[Backtracing] Fix runtime symbols.

The runtime functions must be marked `extern "C"` because the symbols
are not mangled when exported.

rdar://137201928

al45tair commented:
@swift-ci Please test

al45tair committed:
[Backtracing] memserver_req/resp and thread are only for Linux.

These types only exist on Linux, so need to be conditionalised here.

rdar://137201928

al45tair commented:
@swift-ci Please test

al45tair commented:
@swift-ci Please test
